### PR TITLE
K8s app as gp child

### PR DESCRIPTION
**Pull request by dvoet**
Ticket: https://broadworkbench.atlassian.net/browse/CA-1167

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

**Comment by dvoet** [comment]
jenkins retest

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -277,15 +277,21 @@ resourceTypes = {
...
       }
+      add_child = {
+        description = "add child to google project"
+      }
     }
     ownerRoleName = "owner"
     roles = {
       owner = {
-        roleActions = ["read_policies", "list_notebook_cluster", "launch_notebook_cluster", "delete_notebook_cluster", "list_persistent_disk", "create_persistent_disk", "delete_persistent_disk", "create_kubernetes_app", "delete", "get_parent", "set_parent"]
+        roleActions = ["read_policies", "list_notebook_cluster", "delete_notebook_cluster", "list_persistent_disk", "delete_persistent_disk", "delete", "get_parent", "set_parent"]
```
Comment (dvoet) [inline src/main/resources/reference.conf:287]: removed some actions here that are included by `notebook-user`

File: src/main/resources/reference.conf [hunk 2]
```diff
@@ -277,15 +277,21 @@ resourceTypes = {
...
         includedRoles = ["notebook-user"]
+        descendantRoles = {
+          kubernetes-app = ["manager"]
+        }
       }
       notebook-user = {
-        roleActions = ["launch_notebook_cluster", "create_persistent_disk", "create_kubernetes_app"]
+        roleActions = ["launch_notebook_cluster", "create_persistent_disk", "create_kubernetes_app", "add_child"]
```
Comment (dvoet) [inline src/main/resources/reference.conf:294]: `create_kubernetes_app` will require added a child to the google-project

File: src/main/resources/reference.conf [hunk 3]
```diff
@@ -460,11 +466,14 @@ resourceTypes = {
...
       }
+      set_parent = {
+        description = "set parent of kubernetes app"
+      }
     }
     ownerRoleName = "creator"
     roles = {
       creator = {
-        roleActions = ["delete", "connect", "update", "status", "stop", "start", "read_policies"]
+        roleActions = ["delete", "connect", "update", "status", "stop", "start", "read_policies", "set_parent"]
```
Comment (dvoet) [inline src/main/resources/reference.conf:476]: in order to set a `google-project` as parent `creator` (the owner role) needs to have `set_parent`

File: src/main/resources/reference.conf [hunk 4]
```diff
@@ -277,15 +277,21 @@ resourceTypes = {
...
     }
     ownerRoleName = "owner"
     roles = {
       owner = {
-        roleActions = ["read_policies", "list_notebook_cluster", "launch_notebook_cluster", "delete_notebook_cluster", "list_persistent_disk", "create_persistent_disk", "delete_persistent_disk", "create_kubernetes_app", "delete", "get_parent", "set_parent"]
+        roleActions = ["read_policies", "list_notebook_cluster", "delete_notebook_cluster", "list_persistent_disk", "delete_persistent_disk", "delete", "get_parent", "set_parent"]
         includedRoles = ["notebook-user"]
+        descendantRoles = {
+          kubernetes-app = ["manager"]
+        }
```
Comment (dvoet) [inline src/main/resources/reference.conf:291]: this is the magic: the owner `role` of `google-project` confers the `manager` role on `kubernetes-app`s

**Review by marctalbott — APPROVED** [review]

**Comment by rtitle** [comment]
The Leo side is merged, tests should hopefully pass if you jenkins retest

File: src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20201102_flattened_role_materialized_view.xml [hunk 5]
```diff
@@ -5,13 +5,17 @@
...
+            DROP MATERIALIZED VIEW IF EXISTS sam_flattened_role
+        </sql>
         <sql stripComments="true">
             create materialized view sam_flattened_role as (
             with recursive flattened_role(base_role_id, nested_role_id, descendants_only) as (
             select resourceRole.id, resourceRole.id, false from sam_resource_role resourceRole
             union
-            select flattened_role.base_role_id, nestedRole.nested_role_id, nestedRole.descendants_only from sam_nested_role nestedRole
+            -- once a role in the inheritance chain is descendants_only then all lower roles in the chain are descendants_only too
+            select flattened_role.base_role_id, nestedRole.nested_role_id, nestedRole.descendants_only or flattened_role.descendants_only from sam_nested_role nestedRole
```
Comment (andy7i) [inline src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20201102_flattened_role_materialized_view.xml]: I was confused about why all descendants (of a role with descendants_only) also have descendants_only. Marc took some time to try to explain it and came up with this example that helped me: https://broadinstitute.slack.com/archives/CNXF2BAPJ/p1611779010090700?thread_ts=1611774168.086400&cid=CNXF2BAPJ

File: src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20201102_flattened_role_materialized_view.xml [hunk 6]
```diff
@@ -5,13 +5,17 @@
...
+        <sql>
+            DROP MATERIALIZED VIEW IF EXISTS sam_flattened_role
+        </sql>
         <sql stripComments="true">
             create materialized view sam_flattened_role as (
             with recursive flattened_role(base_role_id, nested_role_id, descendants_only) as (
             select resourceRole.id, resourceRole.id, false from sam_resource_role resourceRole
             union
-            select flattened_role.base_role_id, nestedRole.nested_role_id, nestedRole.descendants_only from sam_nested_role nestedRole
+            -- once a role in the inheritance chain is descendants_only then all lower roles in the chain are descendants_only too
```
Comment (andy7i) [inline src/main/resources/org/broadinstitute/dsde/sam/liquibase/changesets/20201102_flattened_role_materialized_view.xml]:
```suggestion
            -- Roles can have "descendant roles" (granted from `descendantRoles`). These roles only apply to the children/descendants of the original resource and not the original resource itself, so descendant roles are marked `descendants_only`. If the descendant role is nested and includes any other roles (`includedRoles`) then logically those roles should also only apply to descendants. Therefore as we unnest/flatten the hierarchy of descendant roles, any additional (included) roles are marked as `descendants_only` as well.
```



Is this accurate? / Too much?

**Review by andy7i — APPROVED** [review]
Do we need to revise the security impact of this ticket since it does affect the DB / permissions?

Added a suggestion for explaining the purpose of the SQL statement. The extra detail will help me a lot the next time I look at this query. Thumbing this PR assuming additional explanation is added.

**Pull request merged**